File: src/org/knuchel/refactorToFunctionnal/talk/v0/KMeans.scala
package org.knuchel.refactorToFunctionnal.talk.v0

import java.util.ArrayList
import java.util.List

import scala.collection.JavaConversions.asScalaBuffer

import org.knuchel.refactorToFunctionnal.talk.beans.Cluster
import org.knuchel.refactorToFunctionnal.talk.beans.Iris
import org.knuchel.refactorToFunctionnal.talk.utils.Util

object KMeans {

  def fitClusters(clusters: List[Cluster], dataset: List[Iris]): List[Cluster] = {
    var testClusters: List[Cluster] = null
    var newClusters = clusters
    while (newClusters != testClusters) {
      testClusters = newClusters
      val clusteredDataset = getClusteredDataset(testClusters, dataset)
      newClusters = improveClusters(clusteredDataset)
    }
    testClusters
  }

  def getClusteredDataset(clusters: List[Cluster], dataset: List[Iris]): List[(Cluster, List[Iris])] = {
    val clusteredIris = new ArrayList[(Cluster, List[Iris])]()
    for (cluster <- clusters) {
      clusteredIris.add((cluster, new ArrayList[Iris]()))
    }
    for (iris <- dataset) {
      val irisCluster = getClusterFor(iris, clusters)
      for (pair <- clusteredIris if pair._1 == irisCluster) {
        pair._2.add(iris)
        //break
      }
    }
    clusteredIris
  }

  def getClusterFor(iris: Iris, clusters: List[Cluster]): Cluster = {
    var nearestCluster: Cluster = null
    var smallestDistance: java.lang.Double = null
    for (cluster <- clusters) {
      val clusterDistance = Util.distance(cluster, iris)
      if (smallestDistance == null || smallestDistance > clusterDistance) {
        smallestDistance = clusterDistance
        nearestCluster = cluster
      }
    }
    nearestCluster
  }

  private def improveClusters(clusteredDataset: List[(Cluster, List[Iris])]): List[Cluster] = {
    val clusters = new ArrayList[Cluster]()
    for (entry <- clusteredDataset) {
      val cluster = entry._1
      val clusterSet = entry._2
      if (clusterSet != null && clusterSet.size > 0) {
        var sepalLengthSum = 0d
        var sepalWidthSum = 0d
        var petalLengthSum = 0d
        var petalWidthSum = 0d
        for (iris <- clusterSet) {
          if (iris.getSepalLength != null) {
            sepalLengthSum += iris.getSepalLength
          }
          if (iris.getSepalWidth != null) {
            sepalWidthSum += iris.getSepalWidth
          }
          if (iris.getPetalLength != null) {
            petalLengthSum += iris.getPetalLength
          }
          if (iris.getPetalWidth != null) {
            petalWidthSum += iris.getPetalWidth
          }
        }
        clusters.add(new Cluster(cluster.getName, sepalLengthSum / clusterSet.size, sepalWidthSum / clusterSet.size,
          petalLengthSum / clusterSet.size, petalWidthSum / clusterSet.size))
      } else {
        clusters.add(Cluster.getRandom(cluster.getName))
      }
    }
    clusters
  }
}

File: src/org/knuchel/refactorToFunctionnal/imperative/KMeans.java
package org.knuchel.refactorToFunctionnal.imperative;

import java.util.ArrayList;
import java.util.List;

import org.knuchel.refactorToFunctionnal.imperative.beans.Cluster;
import org.knuchel.refactorToFunctionnal.imperative.beans.Iris;
import org.knuchel.refactorToFunctionnal.imperative.beans.Pair;
import org.knuchel.refactorToFunctionnal.imperative.utils.Util;

public class KMeans {

	public static List<Cluster> fitClusters(List<Cluster> clusters, List<Iris> dataset) {
		List<Cluster> testClusters = null, newClusters = clusters;
		while (!newClusters.equals(testClusters)) {
			testClusters = newClusters;

			// split dataset in clusters
			List<Pair<Cluster, List<Iris>>> clusteredDataset = getClusteredDataset(testClusters, dataset);

			// calculate new clusters with the splited dataset
			newClusters = improveClusters(clusteredDataset);
		}

		return testClusters;
	}

	public static List<Pair<Cluster, List<Iris>>> getClusteredDataset(List<Cluster> clusters, List<Iris> dataset) {
		List<Pair<Cluster, List<Iris>>> clusteredIris = new ArrayList<Pair<Cluster, List<Iris>>>();
		// create cluster lists
		for (Cluster cluster : clusters) {
			clusteredIris.add(new Pair<Cluster, List<Iris>>(cluster, new ArrayList<Iris>()));
		}

		// add data in corresponding cluster (nearest)
		for (Iris iris : dataset) {
			Cluster irisCluster = getClusterFor(iris, clusters);
			for (Pair<Cluster, List<Iris>> pair : clusteredIris) {
				if (pair.getKey().equals(irisCluster)) {
					pair.getValue().add(iris);
					break;
				}
			}
		}

		return clusteredIris;
	}

	public static Cluster getClusterFor(Iris iris, List<Cluster> clusters) {
		Cluster nearestCluster = null;
		Double smallestDistance = null;
		// return the nearest cluster
		for (Cluster cluster : clusters) {
			Double clusterDistance = Util.distance(cluster, iris);
			if (smallestDistance == null || smallestDistance > clusterDistance) {
				smallestDistance = clusterDistance;
				nearestCluster = cluster;
			}
		}
		return nearestCluster;
	}

	private static List<Cluster> improveClusters(List<Pair<Cluster, List<Iris>>> clusteredDataset) {
		List<Cluster> clusters = new ArrayList<Cluster>();
		// calculate new positions for each cluster
		for (Pair<Cluster, List<Iris>> entry : clusteredDataset) {
			Cluster cluster = entry.getKey();
			List<Iris> clusterSet = entry.getValue();
			if (clusterSet != null && clusterSet.size() > 0) {
				// new cluster position is in the center of the cluster elements
				Double sepalLengthSum = 0d, sepalWidthSum = 0d, petalLengthSum = 0d, petalWidthSum = 0d;
				for (Iris iris : clusterSet) {
					if (iris.getSepalLength() != null) {
						sepalLengthSum += iris.getSepalLength();
					}
					if (iris.getSepalWidth() != null) {
						sepalWidthSum += iris.getSepalWidth();
					}
					if (iris.getPetalLength() != null) {
						petalLengthSum += iris.getPetalLength();
					}
					if (iris.getPetalWidth() != null) {
						petalWidthSum += iris.getPetalWidth();
					}
				}
				clusters.add(new Cluster(cluster.getName(), sepalLengthSum / clusterSet.size(), sepalWidthSum / clusterSet.size(), petalLengthSum
						/ clusterSet.size(), petalWidthSum / clusterSet.size()));
			} else {
				// if cluster has no element, his new position is random
				clusters.add(Cluster.getRandom(cluster.getName()));
			}
		}
		return clusters;
	}
}
